src/evaluators/TryStatement.js

```diff
@@ -11,21 +11,55 @@
 
 import type { Realm } from "../realm.js";
 import type { LexicalEnvironment } from "../environment.js";
-import { AbruptCompletion, ThrowCompletion } from "../completions.js";
-import { UpdateEmpty } from "../methods/index.js";
+import { AbruptCompletion, Completion, PossiblyNormalCompletion, ThrowCompletion } from "../completions.js";
+import { joinEffects, UpdateEmpty } from "../methods/index.js";
 import { Value } from "../values/index.js";
 import type { BabelNodeTryStatement } from "babel-types";
 import invariant from "../invariant.js";
 
-export default function(ast: BabelNodeTryStatement, strictCode: boolean, env: LexicalEnvironment, realm: Realm): Value {
+export default function(
+  ast: BabelNodeTryStatement,
+  strictCode: boolean,
+  env: LexicalEnvironment,
+  realm: Realm
+): PossiblyNormalCompletion | Value {
   let completions = [];
 
-  let blockRes = env.evaluateCompletion(ast.block, strictCode);
-
-  if (blockRes instanceof ThrowCompletion && ast.handler) {
-    completions.unshift(env.evaluateCompletion(ast.handler, strictCode, blockRes));
+  let blockRes = env.evaluateAbstractCompletion(ast.block, strictCode);
+  if (blockRes instanceof PossiblyNormalCompletion) {
+    let abruptCompletion;
+    let abruptEffects;
+    if (blockRes.consequent instanceof AbruptCompletion) {
+      abruptCompletion = blockRes.consequent;
+      abruptEffects = blockRes.consequentEffects;
+    } else {
+      abruptCompletion = blockRes.alternate;
+      abruptEffects = blockRes.alternateEffects;
+    }
+    if (abruptCompletion instanceof ThrowCompletion && ast.handler) {
+      let normalEffects = realm.getCapturedEffects(blockRes.value);
+      invariant(normalEffects !== undefined);
+      realm.stopEffectCaptureAndUndoEffects();
+      let handlerEffects = realm.evaluateForEffects(() => {
+        realm.applyEffects(abruptEffects);
+        invariant(ast.handler);
+        return env.evaluateAbstractCompletion(ast.handler, strictCode, abruptCompletion);
+      });
+      let jointEffects;
+      if (blockRes.consequent instanceof AbruptCompletion)
+        jointEffects = joinEffects(realm, blockRes.joinCondition, handlerEffects, normalEffects);
```

Comment:
Why is the order of the effects important here?

Reply:
Because of the join condition.